dbt/include/global_project/macros/materializations/table/table.sql

```diff
@@ -60,12 +63,17 @@
   {% if non_destructive_mode -%}
     -- noop
   {%- else -%}
-    {{ drop_relation_if_exists(old_relation) }}
+    {% if exists_as_table %}
+      -- move the existing table out of the way
+      {{ adapter.rename_relation(target_relation, backup_relation) }}
+    {% endif %}
+
     {{ adapter.rename_relation(intermediate_relation, target_relation) }}
   {%- endif %}
 
   -- `COMMIT` happens here
   {{ adapter.commit() }}
+  {{ drop_relation_if_exists(backup_relation) }}
```

Comment:
would this logic fail if the backup relation wasn't deleted on the previous run? i.e.:

run 1:
- dbt renames target to backup, commit succeeds, but then it fails to drop backup here
run 2:
- ERROR: relation already exists?

i'd imagine you could fix this by also dropping the backup relation at the start of the run, if it exists.

Reply:
really good point. We do something similar for the __dbt_tmp relation -- i will update!